File: Pregunta2/E2.py
# Fibonnaci base 3 con funciones temporales

suma3 = lambda a, b, c: a + b + c
fibo = lambda n: fiboRec(n, -1, 1, 0)

def fiboRec(n, a, b, c):
    if n >= 0:
        f = suma3(a, b, c)
        print(f, end=" ")
        fiboRec(n-1, b, c, f)

n = int(input("Ingrese numero de terminos: "))
print("fibonacci base 3 con funciones temporales")
fibo(n)

File: Pregunta2/E2.hs
--ejercicio 2
-- resolver el ejericio 1 con funciones temporales
-- fibonnaci base 3 forma recursiva
fiboRec =
    let x 0 = 0
        x 1 = 1
        x 2 = 1
        x n = fiboRec (n-3) + fiboRec (n-2) + fiboRec (n-1)
        fx = map x [0 .. ]
    in \f -> (fx !! f)

-- fibonacci base 3 forma estructurada
fiboEst n = take n . map head $ iterate (\(a:b:c:d) -> b:c:(a+b+c):d) [0,1,1]

main = do 
    putStrLn "Ingrese un numero: "
    n <- getLine
    let num = (read n :: Int)
    putStrLn "fibonacci base 3 forma estructurada: "
    print (fiboEst num)
    putStrLn "fibonacci base 3 forma recursiva "
    print (map fiboRec[0..num-1])
    
